lib/eventFactory.js

```diff
@@ -2,9 +2,116 @@
+ * For example, if startFrom ~commit, then create builds with jobs that have requires: ~commit
+ * @method createBuilds
+ * @param  {Pipeline} pipeline                   Pipeline to create builds
+ * @param  {Object}   config
+ * @param  {String}   config.sha                 SHA this project was built on
+ * @param  {String}   config.username            Username of the user that creates this event
+ * @param  {String}   config.scmContext          The scm context to which user belongs
+ * @param  {String}   [config.prRef]             Ref if it's a PR event
+ * @param  {String}   [config.prNum]             PR Number if it's a PR event
+ * @param  {String}   [config.startFrom]         Where the event starts from (jobname or ~commit, ~pr, etc)
+ *                                               Making optional for backward compatibility
+ * @param  {Number}   eventId                    Event's id
+ */
+function createBuilds(pipeline, config, eventId) {
+    const startFrom = config.startFrom;
+
+    // If no startFrom, then do nothing. Remove once we switch to new workflow design
+    if (!startFrom) {
+        return null;
+    }
+
+    /* eslint-disable global-require */
+    const BuildFactory = require('./buildFactory');
+    const buildFactory = BuildFactory.getInstance();
+    const JobFactory = require('./jobFactory');
+    const jobFactory = JobFactory.getInstance();
+    /* eslint-enable global-require */
+
+    return pipeline.jobs.then((jobs) => {
+        const jobNamesArray = jobs.map(j => j.name);
+        const jobsToCreate = [];
+        let jobsToStart = [];
+
+        jobs.forEach((job) => {
+            const jobName = job.name;
+            const isEnabled = job.state === 'ENABLED';
+            const isJobName = !joi.validate(startFrom, jobConfigSchema.jobname).error;
+            const isTrigger = !joi.validate(startFrom, jobConfigSchema.trigger).error;
+            const requires = job.permutations.requires;
+
+            // Do nothing if job is disabled
+            if (isEnabled) {
+                // If startFrom is a jobname, then start that job
+                if ((isJobName && startFrom === jobName)) {
+                    jobsToStart.push(job);
+                // If startFrom is a trigger (example: ~commit, ~pr), then start jobs that include that key
+                } else if (isTrigger && requires && requires.includes(startFrom)) {
+                    const prJobName = `PR-${config.prNum}-${jobName}`;
+
+                    // If PR job already exists, skip
+                    if (jobNamesArray.includes(prJobName)) {
+                        return;
+                    }
+
+                    // If startFrom is not ~pr, or if ~pr and isPRJob already, start job
+                    if (startFrom !== '~pr' || job.isPR()) {
+                        // Skip if PR number does not match
+                        if (startFrom === '~pr' && job.isPR() && job.prNum() !== config.prNum) {
```

Comment:
Can you move this out to top? There is too much nesting right now. I would have something like 
```
jobs
  .filter(filterUnwantedPRJobs ?)
  .forEach()
  ....
)
```

Reply:
I tried. 😢 🆘 
Can you take a look?